tests/unit/test-util/test.submit-addon.js

```diff
@@ -257,6 +265,133 @@ describe('util.submit-addon', () => {
+          getUploadUuidStub
+        );
+
+        assert.equal(returnedUuid, newUuid);
+        sinon.assert.calledWith(getUploadUuidStub, uuidFilePath);
+        sinon.assert.calledWith(doUploadStub, xpiPath, channel);
+        sinon.assert.calledWith(saveUploadUuidStub, uuidFilePath, {
+          uploadUuid: newUuid,
+          channel,
+          xpiCrcHash: newHash,
+        });
+      });
+
+      it('skips doUploadSubmit if previous hash is the same as current', async () => {
+        const xpiCrcHash = 'some-hash';
+        const uploadUuid = '{some-uuid}';
+        const xpiPath = 'some/path/to/file.xpi';
+        const uuidFilePath = 'some/path/to/.uploadUuid';
+        const channel = 'someChannel';
+        const client = new Client(clientDefaults);
+        sinon.stub(client, 'hashXpiCrcs').resolves(xpiCrcHash);
+        const doUploadStub = sinon.stub(client, 'doUploadSubmit');
+        const saveUploadUuidStub = sinon.stub();
+        const getUploadUuidStub = sinon
+          .stub()
+          .resolves({ uploadUuid, channel, xpiCrcHash });
+
+        const returnedUuid = await client.getPreviousUuidOrUploadXpi(
+          xpiPath,
+          channel,
+          uuidFilePath,
+          saveUploadUuidStub,
+          getUploadUuidStub
+        );
+
+        assert.equal(returnedUuid, uploadUuid);
+        sinon.assert.calledWith(getUploadUuidStub, uuidFilePath);
+        sinon.assert.notCalled(doUploadStub);
+        sinon.assert.notCalled(saveUploadUuidStub);
+      });
+
+      it('calls doUploadSubmit if the channel is different to current', async () => {
+        const xpiCrcHash = 'some-hash';
+        const oldUuid = '{some-uuid}';
+        const newUuid = '{some-other-uuid}';
+        const xpiPath = 'some/path/to/file.xpi';
+        const uuidFilePath = 'some/path/to/.uploadUuid';
+        const oldChannel = 'someChannel';
+        const newChannel = 'someOtherChannel';
+        const client = new Client(clientDefaults);
+        sinon.stub(client, 'hashXpiCrcs').resolves(xpiCrcHash);
+        const doUploadStub = sinon
+          .stub(client, 'doUploadSubmit')
+          .resolves(newUuid);
+        const saveUploadUuidStub = sinon.stub();
+        const getUploadUuidStub = sinon
+          .stub()
+          .resolves({ uploadUuid: oldUuid, channel: oldChannel, xpiCrcHash });
+
+        const returnedUuid = await client.getPreviousUuidOrUploadXpi(
+          xpiPath,
+          newChannel,
+          uuidFilePath,
+          saveUploadUuidStub,
+          getUploadUuidStub
+        );
+
+        assert.equal(returnedUuid, newUuid);
+        sinon.assert.calledWith(getUploadUuidStub, uuidFilePath);
+        sinon.assert.calledWith(doUploadStub, xpiPath, newChannel);
+        sinon.assert.calledWith(saveUploadUuidStub, uuidFilePath, {
+          uploadUuid: newUuid,
+          channel: newChannel,
+          xpiCrcHash,
+        });
+      });
+    });
+
+    describe('hashXpiCrcs', () => {
```

Comment:
Need more test cases, where we don't necessarily care about the output (i.e. CRC32), but that the hashes are different. At least the following (but depending on the implementation of hashXpiCrcs you may need more):
- zip with directory != zip without that directory.
- zip with directory != zip with empty file.
- zip with file A == zip with file a (same content, file name differs)
- zip with file a b == zip with b a (file order differs).

Reply:
> zip with file a b == zip with b a (file order differs).

@Rob--W this case should return a _different_ hash, or the same?  (AMO rewrites the zip file server-side so the file order shouldn't be relied upon by the extension.)

Reply:
Same hash.